code/modules/antagonists/wizard/equipment/spellbook_entries/summons.dm

```diff
@@ -76,12 +76,71 @@
 	cost = 4
 
-/datum/spellbook_entry/summon/curse_of_madness/buy_spell(mob/living/carbon/human/user, obj/item/spellbook/book)
+/datum/spellbook_entry/summon/curse_of_madness/buy_spell(mob/living/carbon/human/user, obj/item/spellbook/book, log_buy = TRUE)
 	var/message = tgui_input_text(user, "Whisper a secret truth to drive your victims to madness", "Whispers of Madness")
-	if(!message)
+	if(!message || QDELETED(user) || QDELETED(book) || !can_buy(user, book))
 		return FALSE
 	curse_of_madness(user, message)
 	playsound(user, 'sound/magic/mandswap.ogg', 50, TRUE)
 	return ..()
 
+/datum/spellbook_entry/summon/specific_spell
+	name = "Mass Wizard Teaching"
+	desc = "Teach a specific spell (or give a specific item) to everyone on the station. \
+		The cost of this is increased by the cost of the spell you choose. And don't worry - you, too, will learn the spell!"
+	cost = 3 // cheapest is 4 cost, most expensive is 7 cost
+	limit = 1
+
+/datum/spellbook_entry/summon/specific_spell/buy_spell(mob/living/carbon/human/user, obj/item/spellbook/book, log_buy = TRUE)
+	var/list/spell_options = list()
+	for(var/datum/spellbook_entry/entry as anything in book.entries)
+		if(istype(entry, /datum/spellbook_entry/summon))
+			continue
+		if(!entry.can_be_purchased())
+			continue
+
+		spell_options[entry.name] = entry
+
+	sortTim(spell_options, GLOBAL_PROC_REF(cmp_text_asc))
+	var/chosen_spell_name = tgui_input_list(user, "Choose a spell (or item) to grant to everyone...", "Wizardly Teaching", spell_options)
+	if(isnull(chosen_spell_name) || QDELETED(user) || QDELETED(book))
+		return FALSE
+	if(GLOB.mass_teaching)
+		tgui_alert(user, "Someone's already cast [name]!", "Wizardly Teaching", list("Shame"))
+		return FALSE
+
+	var/datum/spellbook_entry/chosen_entry = spell_options[chosen_spell_name]
+	if(cost + chosen_entry.cost > book.uses)
+		tgui_alert(user, "You can't afford to grant everyone [chosen_spell_name]! ([cost] points needed)", "Wizardly Teaching", list("Shame"))
+		return FALSE
+
+	cost += chosen_entry.cost
+	if(!can_buy(user, book))
+		cost = initial(cost)
+		return FALSE
+
+	GLOB.mass_teaching = new(chosen_entry.type)
+	GLOB.mass_teaching.equip_all_affected()
+
+	var/item_entry = istype(chosen_entry, /datum/spellbook_entry/item)
+	to_chat(user, span_hypnophrase("You have [item_entry ? "granted everyone the power" : "taught everyone the ways"] of [chosen_spell_name]!"))
+	message_admins("[ADMIN_LOOKUPFLW(user)] gave everyone the [item_entry ? "item" : "spell"] \"[chosen_spell_name]\"!")
+	user.log_message("has gave everyone the [item_entry ? "item" : "spell"] \"[chosen_spell_name]\"!", LOG_GAME)
+
+	name = "[name]: [chosen_spell_name]"
+	return ..()
+
+/datum/spellbook_entry/summon/specific_spell/can_buy(mob/living/carbon/human/user, obj/item/spellbook/book)
+	if(GLOB.mass_teaching)
+		return FALSE
```

Comment:
You already check this at L108? Or does this do something else aswell?

Reply:
This check is for general sanity, such as buying it from the spellbook, while the one at 108 is separated out to give specific feedback if someone else cast it while they were in the input menu